### PR TITLE
Smart jdbc

### PR DESCRIPTION
**Pull request by kneeraj**
Done the following changes:

- Added load balancing capabilities in the jdbc driver
- Modified the YugabyteClusterAwareDriver. It extends PGSimpleDataSource
- Added 2 new properties: _load-balance_ and _topology-keys_
- when only load-balance=true is given then uniform load distribution happens and when topolgy-keys are given then connections are distributed to only those hosts.
- removed cassandra dependencies and older classes.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
3. [ ] Have you added your new test classes to an existing test suite?

### Changes to Existing Features:

* [x] Does this break existing behaviour? If so please explain.
* [x ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x ] Have you written new tests for your core changes, as applicable?
* [x ] Have you successfully run tests with your changes locally?

### CODE REVIEW

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    LOGGER.log(Level.FINE, "Getting the list of servers");
+    ResultSet rs = st.executeQuery(GET_SERVERS_QUERY);
+    ArrayList<String> currentServers = new ArrayList<>();
+    while (rs.next()) {
+      String host = rs.getString("host");
+      currentServers.add(host);
+    }
+    LOGGER.log(Level.FINE, "List of servers got {0}", currentServers);
+    // System.out.println("List of servers got: " + currentServers);
+    // System.out.println();
```
Comment (m-iancu) [inline jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:73]: remove these two commented out lines?

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/TopologyAwareLoadBalancer.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        } else if (set == null) {
+          HashSet<String> zoneset = new HashSet<>();
+          zoneset.add(regionandzone[1]);
+          allowedPlacements.put(regionandzone[0], zoneset);
+        }
+      }
+    }
+  }
+
+  protected ArrayList<String> getCurrentServers(Connection conn) throws SQLException {
```
Comment (m-iancu) [inline jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/TopologyAwareLoadBalancer.java:46]: should this have an `@Override` annotation?

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/TopologyAwareLoadBalancer.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    placements = placementvalues;
+    allowedPlacements = new HashMap<>();
+    populateMap();
+  }
+
+  protected String loadBalancingNodes() {
+    return placements;
+  }
+
+  private void populateMap() {
```
Comment (m-iancu) [inline jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/TopologyAwareLoadBalancer.java:26]: nit: maybe `populatePlacementMap` is a clearer name here.

File: jdbc-yugabytedb/src/main/java/org/postgresql/ds/common/BaseDataSource.java [hunk 4]
```diff
@@ -1183,7 +1192,9 @@ public void setProperty(PGProperty property, String value) {
         break;
       case PG_PORT:
         try {
-          portNumber = Integer.parseInt(value);
+          if (portNumber == 0) {
+            portNumber = Integer.parseInt(value);
+          }
```
Comment (m-iancu) [inline jdbc-yugabytedb/src/main/java/org/postgresql/ds/common/BaseDataSource.java:1197]: what is the overriding rule here? For the other cases it looks like we're not checking the previous value but just overriding whatever was there before. Why are we differing for port (i.e. only setting it if it is `0`/unset previously)?

File: examples/src/main/java/com/yugabyte/examples/TopologyAwareLoadBalanceExample.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        + ":" + controlPort + "/yugabyte?user=yugabyte&password=yugabyte&load-balance=true&topology-keys=region1.zone1";
+
+      System.out.println("Setting up the connection pool having 6 connections.......");
+
+      Thread.sleep(1000);
+
+      testUsingHikariPool("topology_aware_load_balance", "region1.zone1", "region1.zone1",
+        controlHost, controlPort, numConnections, VERBOSE, INTERACTIVE);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
```
Comment (tedyu) [inline examples/src/main/java/com/yugabyte/examples/TopologyAwareLoadBalanceExample.java:38]: Please restore Interrupt Status Flag.

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
+package com.yugabyte.ysql;
+
+import java.sql.*;
+import java.util.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class ClusterAwareLoadBalancer {
+  protected static final String GET_SERVERS_QUERY = "select * from yb_servers()";
+  protected static final Logger LOGGER = Logger.getLogger("org.postgresql.Driver");
```
Comment (tedyu) [inline jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:10]: It seems the logger name should reflect current class.

File: jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java [hunk 7]
```diff
@@ -450,12 +451,89 @@ public Connection getResult(long timeout) throws SQLException {
...
+        loadBalancer.refresh(controlConnection);
+        controlConnection.close();
+      } catch (PSQLException ex) {
+        if (PSQLState.UNDEFINED_FUNCTION.getState().equals(ex.getSQLState())) {
+          return null;
+        }
+        throw ex;
+      }
+    }
+    chosenHost = loadBalancer.getLeastLoadedServer(failedHosts);
```
Comment (tedyu) [inline jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:500]: This seems to be the same call as that on line 480.
why is this needed ?

Comment (tedyu) [reply jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:500]: I think this line should be executed when chosenHost was null.
The expectation is that, after refreshing load balancer, we can pick the least loaded server.
Maybe move this call inside the if block after line 498.

File: examples/pom.xml [hunk 8]
```diff
@@ -31,7 +39,7 @@
         <configuration>
           <transformers>
             <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
-              <mainClass>com.yugabyte.examples.Main</mainClass>
+<!--              <mainClass>com.yugabyte.examples.Main</mainClass>-->
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply examples/pom.xml:42]: this Main class was a part of the previous example, we can remove the Main.java class and this comment also. I will do it.

File: examples/src/main/java/com/yugabyte/examples/TopologyAwareLoadBalanceExample.java [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    }
+  }
+
+  private static void testUsingHikariPool(String poolName, String lbpropvalue, String lookupKey,
+                                          String hostName, String port, String numConnections, Boolean VERBOSE, Boolean INTERACTIVE) {
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply examples/src/main/java/com/yugabyte/examples/TopologyAwareLoadBalanceExample.java:43]: Nothing specific it was just at that moment, I will make change this to normal format. Thanks

File: jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java (hunk 7 above)
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:500]: Yes. You are right. Thanks.

File: jdbc-yugabytedb/src/main/java/org/postgresql/ds/common/BaseDataSource.java (hunk 4 above)
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/org/postgresql/ds/common/BaseDataSource.java:1197]: There is a separate addPort method in the datasource. We are giving priority to that if set explicitly.

**Review by kneeraj — COMMENTED** [review]
Replied on the comments made in driver's code. Following it up shortly with incorporating the suggestions.

File: examples/src/main/java/com/yugabyte/examples/TopologyAwareLoadBalanceExample.java [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        System.out.println("You can verify the connections on the server side using your browser");
+        System.out.println("For example, you can visit \"127.0.0.1:13000/rpcz\"" + " and similarly for others...");
+      }
+
+      continueScript("flag3");
+      pauseApp(".jdbc_example_app_checker3");
+
+      System.out.println("Closing the java app...");
+      hikariDataSource.close();
+      Thread.sleep(1000);
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply examples/src/main/java/com/yugabyte/examples/TopologyAwareLoadBalanceExample.java:107]: I put it during initially for some debugging i think. Will remove this now, not needed.

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+  public synchronized void updateConnectionMap(String host, int incDec) {
+    LOGGER.log(Level.FINE, "updating connection count for {0} by {1}",
+      new String[]{host, String.valueOf(incDec)});
+    Integer currentCount = hostToNumConnMap.get(host);
+    if (currentCount == 0 && incDec < 0) return;
+    if (currentCount == null && incDec > 0) {
+      hostToNumConnMap.put(host, incDec);
+    } else if (currentCount != null ) {
+      hostToNumConnMap.put(host, currentCount + incDec);
```
Comment (tedyu) [inline jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:104]: Should we check the case currentCount + incDec < 0 ?
Condition on line 100 can be modified to cover the case.

File: jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java [hunk 12]
```diff
@@ -450,12 +451,89 @@ public Connection getResult(long timeout) throws SQLException {
...
    */
-  private static Connection makeConnection(String url, Properties props) throws SQLException {
-    return new PgConnection(hostSpecs(props), user(props), database(props), props, url);
+  private static Connection makeConnection(String url, Properties properties) throws SQLException {
+    LoadBalanceProperties lbprops = new LoadBalanceProperties(url, properties);
+    if (lbprops.hasLoadBalance()) {
+      Connection conn = getConnectionBalanced(lbprops);
+      if (conn != null) {
+        return conn;
+      } else {
```
Comment (tedyu) [inline jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:464]: the 'else' keyword is not needed (since the if block would return)

File: jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java [hunk 13]
```diff
@@ -450,12 +451,89 @@ public Connection getResult(long timeout) throws SQLException {
...
+        LOGGER.log(Level.WARNING, "Failed to apply load balance. Trying normal connection");
+        // but remember to use stripped url and props
+        url = lbprops.getStrippedURL();
+        properties = lbprops.getStrippedProperties();
+      }
+    }
+    return new PgConnection(hostSpecs(properties), user(properties), database(properties), properties, url);
+  }
+
+  private static Connection getConnectionBalanced(LoadBalanceProperties lbprops) throws SQLException {
```
Comment (tedyu) [inline jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:474]: getConnectionBalanced -> getBalancedConnection

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+  public List<String> getServers() {
+    return this.servers;
+  }
+
+  public synchronized void updateConnectionMap(String host, int incDec) {
+    LOGGER.log(Level.FINE, "updating connection count for {0} by {1}",
+      new String[]{host, String.valueOf(incDec)});
+    Integer currentCount = hostToNumConnMap.get(host);
+    if (currentCount == 0 && incDec < 0) return;
```
[the remark this answers is not in the document]
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:98]: well this is not exactly unexpected. This comes in the picture because the first connection which is made to obtain the server information is closed without making an entry into the internal accounting map. Therefore this special handling.

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/TopologyAwareLoadBalancer.java (hunk 2 above)
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/TopologyAwareLoadBalancer.java:45]: done

File: jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java (hunk 13 above)
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:474]: No, it is deliberately named like that. There is a getConnection method. I have introduced a 'balanced' variant of that. If you keep it 'getBalancedConnection' it will sound more like an attribute of the connection which is not the case here.

File: jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java (hunk 12 above)
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:464]: It may return null if it is talking to servers which are not upgraded with the latest added yb_servers() function but still the client is trying to use the load-balanced property. It is simply warning it that 'load-balanced' attempt  failed and then it fails back to the original behaviour. It's a guard against client server incompatibility for load-balance feature.

Comment (tedyu) [reply jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java:464]: I didn't say else block is not needed.
You can keep the block, omitting the 'else'

This is minor, though.

Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java]: done

**Comment by tedyu** [comment]
LGTM

File: jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java [hunk 15]
```diff
@@ -59,24 +63,46 @@ public boolean needsRefresh() {
...
+        useHostColumn = Boolean.FALSE;
+      }
+    }
+    return getPrivateOrPublicServers(useHostColumn, currentPrivateIps, currentPublicIps);
+  }
+
+  protected ArrayList<String> getPrivateOrPublicServers(Boolean useHostColumn, ArrayList<String> privateHosts, ArrayList<String> publicHosts) {
+    if (useHostColumn == null) {
+      LOGGER.log(Level.WARNING, "Either private or public should have been determined");
+      return null;
```
[the remark this answers is not in the document]
Comment (kneeraj) [reply jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:97]: null is being handled as the refresh 'failed' in which case it falls back to acquiring a non-balanced connection. Let me know if you see a gap here after looking at the code.

[the remark this answers is not in the document]
Comment (m-iancu) [reply jdbc-yugabytedb/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:97]: Got it @kneeraj -- sounds good,  LGTM!
